jax/interpreters/pxla.py

```diff
@@ -412,41 +412,94 @@ def _shard_abstract_array(size, axis: int, x):
   return x.update(shape=tuple_delete(x.shape, axis))
 shard_aval_handlers[ShapedArray] = _shard_abstract_array
 
-def aval_to_result_handler(sharding_spec: Optional[ShardingSpec],
-                           indices: Optional[Tuple[Index]],
-                           aval: core.AbstractValue) -> Callable[
-                               [List[xb.xla_client.Buffer]], Any]:
+MeshAxisName = Any
+"""
+ArrayMapping specifies how an ndarray should map to mesh axes.
+
+Note that the ordering is crucial for the cases when this mapping is non-injective
+(i.e. when multiple mesh axes map to the same positional axis). Then, the
+order of entries of the mapping determines a major-to-minor order on mesh axes,
+according to which chunks of the value along the repeated dimension will be assigned.
+
+For example, consider a mapping {'x': 1, 'y': 1} and a mesh with shape {'x': 2, 'y': 3}.
+The second dimension of the value would get chunked into 6 pieces, and assigned to the
+mesh in a way that treats 'y' as the fastest changing (minor) dimension. In this case,
+that would mean that a flat list of chunks would get assigned to a flattened list of
+mesh devices without any modifications. If the mapping was {'y': 1, 'x': 1}, then the
+mesh devices ndarray would have to be transposed before flattening and assignment.
+"""
+ArrayMapping = OrderedDictType[MeshAxisName, int]
+
+AxisResource = Tuple[Optional[Tuple[Any, ...]], ...]
```

Comment:
What are axis resources in here? If they're supposed to be something like normalized partition specs then why do we allow optionals instead of having a more normalized representation such as `Tuple[Tuple[MeshAxisName, ...], ...]`?

Reply:
I think this should be `Tuple[Union[MeshAxisName, Tuple[MeshAxisName], None], ...]`.
The outer tuple maps to each data dim (does it have to map every dim?). Each element is `MeshAxisName` / `Tuple[MeshAxisName]` to partition that data dim, or `None` if it's unpartitioned. Does that sound right?